homeassistant/components/ps4/config_flow.py

```diff
@@ -99,6 +112,18 @@ def __init__(self):
                     CONF_REGION: self.region
                 }
 
+                if self.hass.config_entries.async_entries(DOMAIN):
+                    # Add new device to previous entry data
+                    _entry.data['devices'].append(device)
+                    # Delete previous entry
+                    await self.hass.config_entries.async_remove(
```

Comment:
Isn't it enough to update the entry? Why do we need to delete the old entry and create a new entry?

Reply:
How do you finish the flow without returning async_create_entry? Wouldn't an exception be raised if a step does not return: abort, form, create?

Reply:
@balloob :point_up:

Do we need another flow result? Or is delete + create the correct approach?

Reply:
@MartinHjelmare So what options are available here? Due to the device api requiring a generated and expiring pin, each device must be configured using data entry flow. I see a few options:

Create a separate entry for each device, restating similar entry data however.
All devices be configured using one entry.
New result for data entry flow to only update existing entry.

Options 1 and 2 can definitely be implemented now. What do you think is most appropriate?

Reply:
If we need to make a separate connection with each device, I think we can make one entry per device.

The connection details will differ between devices and we need to verify the connection per device before creating the entry.

Reply:
Look at the hue config flow. It has a similar approach, although there each host needs to be linked by the user.

Reply:
Hmm. Correct me if I'm wrong but glancing at the Hue flow it handles multiple entries similarly to the my current flow:

```
    async def _entry_from_bridge(self, bridge):
        """Return a config entry from an initialized bridge."""
        # Remove all other entries of hubs with same ID or host
        host = bridge.host
        bridge_id = bridge.config.bridgeid

        same_hub_entries = [entry.entry_id for entry
                            in self.hass.config_entries.async_entries(DOMAIN)
                            if entry.data['bridge_id'] == bridge_id or
                            entry.data['host'] == host]

        if same_hub_entries:
            await asyncio.wait([self.hass.config_entries.async_remove(entry_id)
                                for entry_id in same_hub_entries])

        return self.async_create_entry(
            title=bridge.config.name,
            data={
                'host': host,
                'bridge_id': bridge_id,
                'username': bridge.username,
            }
        )
```

Reply:
If so, this seems to be a valid method correct?

Reply:
> If we need to make a separate connection with each device, I think we can make one entry per device.
> 
> The connection details will differ between devices and we need to verify the connection per device before creating the entry.

This part verifies connection. If login fails then it throws an error message.
```
            is_ready, is_login = await self.hass.async_add_executor_job(
                self.helper.link, self.host, self.creds, self.pin)

            if is_ready is False:
                errors['base'] = 'not_ready'
            elif is_login is False:
                errors['base'] = 'login_failed'
```

Reply:
Hue has one entry per host. It needs to remove entries for same host. In our case, we were trying to have one entry with multiple hosts. Not the same. It's fine to follow the hue example.

Reply:
Ok so I can use update, then create?

Reply:
Why doesn't it make sense to have one entry per device? Don't we need to test the connection to each device before adding it to an entry anyway?

Reply:
Well the reasoning I'm using from a UI standpoint is if looking at the integrations tab you would see two entries: 
PlayStation 4: PlayStation 4
'entity icon'
PlayStation 4: PlayStation 4
'entity icon 2'

Now take Cast for example you see:
Google Cast: Google Cast
'entity icon' 'entity icon 2'

Same goes with components such as Nest and Z-Wave. Delete + create will also result in this. I think that it looks a lot neater on the frontend this way.

From a data point of view the config_entries file would look like this:
```
{
    "data": {
        "entries": [
            {
                "connection_class": "local_poll",
                "data": {
                    "devices": [
                        {
                            "host": "192.168.86.10",
                            "name": "PlayStation 4",
                            "region": "R1"
                        }
                    ],
                    "token": "111111111111111"
                },
                "domain": "ps4",
                "entry_id": "11111111111111111111111111",
                "source": "user",
                "title": "PlayStation 4",
                "version": 1
            },
            {
                "connection_class": "local_poll",
                "data": {
                    "devices": [
                        {
                            "host": "192.168.86.11",
                            "name": "PlayStation 4",
                            "region": "R1"
                        }
                    ],
                    "token": "111111111111111"
                },
                "domain": "ps4",
                "entry_id": "11111111111111111111111112",
                "source": "user",
                "title": "PlayStation 4",
                "version": 1
            }
        ]
    },
    "key": "core.config_entries",
    "version": 1
}
```
compared to:
```
{
    "data": {
        "entries": [
            {
                "connection_class": "local_poll",
                "data": {
                    "devices": [
                        {
                            "host": "192.168.86.10",
                            "name": "PlayStation 4",
                            "region": "R1"
                        },
                        {
                            "host": "192.168.86.11",
                            "name": "PlayStation 4",
                            "region": "R1"
                        }
                    ],
                    "token": "111111111111111"
                },
                "domain": "ps4",
                "entry_id": "11111111111111111111111111",
                "source": "user",
                "title": "PlayStation 4",
                "version": 1
            },

        ]
    },
    "key": "core.config_entries",
    "version": 1
}
```

We test the connection to the device being added. The previous devices would be assumed to have a valid connection since they were configured before, therefore tested already.

Although both have similar end results, I think that delete + create is more concise. Code is also simpler.

Reply:
The credential being the same for all devices speaks for a single config entry. But it could still be simpler to keep one entry per device.

I'd like a second opinion from another member before we continue.

@balloob or @Kane610

Reply:
How would `config_entries.async_foward_entry_setup()`  work with two entries? The first line in the method is `if component not in self.hass.config.components:`  Component is setup from the first entry already though.

Reply:
How do you mean?

Reply:
Do both entries get forwarded to the platform when called?

Reply:
One entry is setup and possibly forwarded at a time. We forward from `async_setup_entry`, which is passed a single entry.

Reply:
Is there no way to get playstations auto discovered?

Reply:
Oh ok. So nothing has to be done on the platform side?

Reply:
@Kane610 Yes its possible.

Reply:
So is it possible to auto add them as well? Or interaction is required? You could keep with one config entry per account that way?

Reply:
No that is not possible. Each PS4 requires the client to acknowledge and send a PIN that it generates and displays for 3 minutes.

Reply:
Ok, so due to the nature of this I'd say that it is more intuitive to create a new config entry per console. You could possibly be smart in the backend and only use one instance per account perhaps

Reply:
Ok sounds good.